File: CrackingTheCodingInterview/Udemy/BackspaceCompare.js
var backspaceCompare = function(s, t) {
    sArr = []
    tArr = []
    for (let i = 0; i < s.length; i++) {
        if (s[i] == '#') {
            sArr.pop()
        } else {
            sArr.push(s[i])
        }
    }
    for (let i = 0; i < t.length; i++) {
        if (t[i] == '#') {
            tArr.pop()
        } else {
            tArr.push(t[i])
        }
    }
    if (sArr.length != tArr.length) {
        return false;
    } else {
        for (let i = 0; i < tArr.length; i++) {
            if (tArr[i] != sArr[i]) {
                return false;
            }
        }
    }
    return true
}

var backspaceCompare2 = function(s, t) {
    let i = s.length-1;
    let j = t.length-1;
    
    while (i >= 0 || j >= 0) {
        if (s[i] === '#' || t[j] === '#') {
            if (s[i] === '#') {
                let backCount = 2;
                while (backCount > 0) {
                    i--;
                    backCount--;
                    if (s[i] === '#') {
                        backCount+=2;
                    }
                }
            }
            if (t[j] === '#') {
                let backCount = 2;
                while (backCount > 0) {
                    j--;
                    backCount--;
                    if (t[j] === '#') {
                        backCount+=2;
                    }
                }
            }
        } else {
            if (s[i] != t[j]) {
                return false;
            } else {
                i--;
                j--;
            }
        }
    }
    return true;
};

File: CrackingTheCodingInterview/Udemy/BackspaceCompare.java
package CrackingTheCodingInterview.Udemy;

import java.util.Stack;

public class BackspaceCompare {
    public static void main(String[] args) {
        
    }

    public boolean backspaceCompare(String s, String t) {
        Stack<Character> sStack = new Stack<>();
        Stack<Character> tStack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '#') {
                if (!sStack.isEmpty()) {
                    sStack.pop();
                }
            } else {
                sStack.push(s.charAt(i));
            }
        }
        for (int i = 0; i < t.length(); i++) {
            if (t.charAt(i) == '#') {
                if (!tStack.isEmpty()) {
                    tStack.pop();
                }
            } else {
                tStack.push(t.charAt(i));
            }
        }
        if (sStack.size() != tStack.size()) {
            return false;
        } else {
            while (!sStack.isEmpty()) {
                if (sStack.pop() != tStack.pop()) {
                    return false;
                }
            }
        }
        return true;
    }

    public boolean backspaceCompare2(String s, String t) {
        int i = s.length()-1;
        int j = t.length()-1;

        while (i >= 0 || j >= 0) {
            if (s.charAt(i) == '#' || t.charAt(j) == '#') {
                if (s.charAt(i) == '#') {
                    int backCount = 2;
                    while (backCount > 2) {
                        i--;
                        backCount--;
                        if (s.charAt(i) == '#') {
                            backCount+=2;
                        }
                    }
                }
                if (t.charAt(j) == '#') {
                    int backCount = 2;
                    while (backCount > 0) {
                        j--;
                        backCount--;
                        if (t.charAt(j) == '#') {
                            backCount+=2;
                        }
                    }
                }
            } else {
                if (s.charAt(i) != t.charAt(j)) {
                    
                }
            }
        }
    }
}